### PR TITLE
fix(jupyter): mark language as `typescriptreact`

### LINKED ISSUE #21079: [jupyter][VS Code] Syntax highlighting and LSP for JSX in Deno Notebooks
Is there something on the Deno LSP side or on the VS Code side for picking up that the notebooks can have JSX in them?

image

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
The Deno kernel already parses and runs code blocks as TSX and not as plain TS (maybe since https://github.com/denoland/deno/pull/20695?). This changes makes vscode use also parse the code blocks as TSX rather than TS.

I tested this locally by manually editing my `~/.local/share/jupyter/kernels/deno/kernel.json`.

Marking as draft because I made this change from the GitHub web editor so I don't know yet if I'll need to update any test, I'll wait for CI to tell me.

Fixes https://github.com/denoland/deno/issues/21079

### CODE REVIEW

File: cli/tools/jupyter/install.rs [hunk 1]
```diff
@@ -99,7 +99,7 @@ pub fn install(
   let json_data = json!({
       "argv": [current_exe_path, "jupyter", "--kernel", "--conn", "{connection_file}"],
       "display_name": display_name,
-      "language": "typescript",
+      "language": "typescriptreact",
```
[the remark this answers is not in the document]
Comment (nicolo-ribaudo) [reply cli/tools/jupyter/install.rs:166]: Source? The Jupyter docs don't mention a restricted list of values.

**Comment by nicolo-ribaudo** [comment]
So, this is a breaking change because existing notebooks have a different language specifier. VSCode is doing some magic and just upgrading them from `typescript` to `typescriptreact`, but other Jupyter clients might complain about the missing kernel.

To avoid that problem, now `deno jupyter --install` will create by default two kernels: one called `Deno` with the language set to `typescriptreact`, and one of legacy compatibility called `Deno (no TSX)` with the language set to `typescript`. It will only create one (with language set to `typescriptreact`) when passing the `--name` or `--display` flags.

**Comment by rgbkrk** [comment]
I think we'd also want to change the KernelInfo message / payload that gets sent over the base protocol.

**Review by fibibot — COMMENTED** [review]
Substance is the right intent for the original report — Deno's Jupyter kernel parses code blocks as TSX (since #20695), so the kernelspec's `language` field should advertise `typescriptreact` so VS Code syntax-highlights cells consistently with what's actually executed. But the implementation adds a **second installed kernel** (`deno-no-tsx`) that has UX implications worth a maintainer's call.

## What I verified

**Refactor + main change** (`cli/tools/jupyter/install.rs:65-184`):

- `install()` extracted into `prepare_kernel_spec_dir(...)` + `install_kernel_spec(...)` helpers (clean refactor, no behavior change in the hot path)
- Primary kernel `kernels/deno/kernel.json` now writes `"language": "typescriptreact"`
- Additionally installs a sibling kernel `kernels/deno-no-tsx/kernel.json` with `"language": "typescript"`, gated on `maybe_name.is_none() && maybe_display_name.is_none()` (so custom-named installs only get the primary)

**Re: @nicolo-ribaudo + @coderabbitai's exchange on validity**: confirmed `typescriptreact` is the standard VS Code language ID for `.tsx` files. Jupyter's kernelspec spec doesn't restrict the `language` field to a fixed enum — VS Code routes it to the language-server provider registered for that ID. Both `typescript` and `typescriptreact` are valid; the latter is what the TS kernel's parsing actually does. ✓

## Concerns

1. **Two-kernel installation by default** — after this PR, `deno jupyter --install` creates *two* entries that both show up in Jupyter's kernel picker: "Deno" (TSX) and "Deno (no TSX)" (plain TS). Users who only want the primary now see two options and may pick the wrong one. Worth weighing whether the no-TSX variant should be opt-in via a flag (e.g., `--also-install-no-tsx`) instead of installed alongside.

2. **`force: false` interaction with the second kernel**: `prepare_kernel_spec_dir` already errors out when the spec directory exists without `force=true`. With this PR, `deno jupyter --install` runs `prepare_kernel_spec_dir` for *both* paths. A user who has the old single-kernel install will hit the error on the primary path; if they pass `--force`, both get overwritten (fine). But if a user manually created a `deno-no-tsx` directory for any reason (maybe they had this PR locally?), they'd hit a confusing error on that path. Minor edge case but worth being defensive.

3. **No tests** — the PR's own description says "Marking as draft because I made this change from the GitHub web editor so I don't know yet if I'll need to update any test". A spec test exercising `deno jupyter --install` and asserting the kernel.json content (especially the new `language: typescriptreact`) would lock in the fix.

4. **CI staleness**: 1 failure (`test debug linux-aarch64`) recorded but logs are aged out (HTTP 410); the CI matrix from December 2025 is also smaller than today's. Needs a rebase + re-run on current main to confirm.

## Escalation

cc @bartlomieju on the two-kernel UX decision — even a small Jupyter-UX change has visible knock-on effects for users who have the kernel installed.

Substance is sound for the language identifier change; just the bundled second-kernel needs a maintainer's read.
